strawberry/tools/depth_limit_validator.py

```diff
@@ -0,0 +1,171 @@
+import re
+from typing import Callable, List, Union
+
+from graphql import GraphQLError
+from graphql.language import (
+    DefinitionNode,
+    FieldNode,
+    FragmentDefinitionNode,
+    FragmentSpreadNode,
+    InlineFragmentNode,
+    OperationDefinitionNode,
+)
+from graphql.validation import ValidationContext, ValidationRule
+
+
+def depth_limit_validator(max_depth: int, options=None, callback=None):
+    """
+    Creates a validator for the GraphQL query depth
+
+    This is a Python port of https://github.com/stems/graphql-depth-limit
+
+    - max_depth - The maximum allowed depth for any operation in a GraphQL document.
+    - options
+        - options.ignore - Stops recursive depth checking based on a field name.
+        Either a string or regexp to match the name, or a function that returns
+        a boolean.
+    - callback - Called each time validation runs. Receives an Object which is a
+    map of the depths for each operation.
+    """
+    if not options:
+        options = {}
+
+    class DepthLimitValidator(ValidationRule):
+        def __init__(self, validation_context: ValidationContext):
+            document = validation_context.document
+            definitions = document.definitions
+
+            fragments = get_fragments(definitions)
+            queries = get_queries_and_mutations(definitions)
+            query_depths = {}
+
+            for name in queries:
+                query_depths[name] = determine_depth(
+                    queries[name],
+                    fragments,
+                    0,
+                    max_depth,
+                    validation_context,
+                    name,
+                    options,
+                )
+
+            if callable(callback):
+                callback(query_depths)
+            super().__init__(validation_context)
+
+    return DepthLimitValidator
+
+
+def get_fragments(definitions: List[DefinitionNode]):
+    fragments = {}
+    for definition in definitions:
+        if isinstance(definition, FragmentDefinitionNode):
+            fragments[definition.name.value] = definition
+
+    return fragments
+
+
+# This will actually get both queries and mutations.
+# We can basically treat those the same
+def get_queries_and_mutations(definitions: List[DefinitionNode]):
+    operations = {}
+
+    for definition in definitions:
+        if isinstance(definition, OperationDefinitionNode):
+            operations[
+                definition.name.value if definition.name else "anonymous"
+            ] = definition
+
+    return operations
+
+
+introspection_regex = re.compile(r"^__")
+
+
+def determine_depth(
+    node, fragments, depth_so_far, max_depth, context, operation_name, options
```

Comment:
Maybe we want to add type hints for these as well? Not sure if necessary, since it's an internal API.